File: monthly_challenges/jan_2021/18.js
// https://leetcode.com/explore/challenge/card/january-leetcoding-challenge-2021/581/week-3-january-15th-january-21st/3608/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxOperations = function(nums, k) {
    const pairs = {};
    let count = 0;
    
    let i = 0;
    while (i < nums.length) {
        num = nums[i];
        // console.log(num, pairs);
        // console.log(`count: ${count}`)
        if (num in pairs) {
            arr = pairs[num]
            if (arr.length > 1) {
                arr.shift();
            } else {
                delete pairs[num] // delete the key in object if there is only one occurrence
            }
            count += 1;
        } else {
            if (k - num in pairs) {
                pairs[k - num].push(i)
            } else {
                pairs[k - num] = [i];
            }
        }
        i += 1;
    }
    
    return count;
};

File: monthly_challenges/feb_2021/18.py
# https://leetcode.com/explore/challenge/card/february-leetcoding-challenge-2021/586/week-3-february-15th-february-21st/3644/

class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        if len(A) < 3:
            return 0

        last = A[1] - A[0]
        digit_streak = 2
        count = 0
        for i in range(2, len(A)):
            num = A[i]
            prev = A[i-1]
            if num - prev == last:
                digit_streak += 1
                count += (digit_streak - 2)
            else:
                digit_streak = 2
                last = num - prev

        return count

"""
Input: [1,2,3,4]
Output: 3

Input: [1,2,3,4,5]
Output: 6

Input: [1,2,3,4,5,6]
Output: 10

[1,4,5,6,7]
Output: 3
"""
